pkg/api/v1/defaults.go

```diff
@@ -389,3 +389,31 @@ func SetDefaults_HostPathVolumeSource(obj *v1.HostPathVolumeSource) {
 		obj.Type = &typeVol
 	}
 }
+
+func SetDefaults_SecurityContext(sc *v1.SecurityContext) {
+	ape := true
+	if sc == nil {
+		sc = &v1.SecurityContext{
+			AllowPrivilegeEscalation: &ape,
+		}
+		return
+	}
+
+	if sc.Privileged != nil && *sc.Privileged {
+		// handle the case for AllowPrivilegeEscalation when the container is privileged
+		sc.AllowPrivilegeEscalation = &ape
+	} else if sc.Capabilities != nil {
+		// handle the case for AllowPrivilegeEscalation when we are adding CAP_SYS_ADMIN
+		for _, cap := range sc.Capabilities.Add {
+			if string(cap) == "CAP_SYS_ADMIN" {
+				sc.AllowPrivilegeEscalation = &ape
+				break
+			}
+		}
+	}
+
+	// handle the case where the user did not set the defaultAllowPrivilegeEscalation and did not explicitly set allowPrivilegeEscalation
+	if sc.AllowPrivilegeEscalation == nil {
+		sc.AllowPrivilegeEscalation = &ape
```

Comment:
this defaulting happens before the pod spec is seen by the PSP admission plugin... if we always default nil values to &true here, the PSP DefaultAllowPrivilegeEscalation will never take effect

Reply:
the fact that we want to be able to set the default value in PSP admission means that we *cannot* set a default via the API defaulting mechanism.

Reply:
oh then we shouldn't even merge this and maybe just keep it the way it is

Reply:
cc @timstclair

Reply:
Good point re:PSP. Here's my new proposal:

- Do the validation (privileged & CAP_SYS_ADMIN) in the API validation
- Don't default in the API defaults
- In the kubelet, just use the reported value from the API, or default to `true`. I.e.:
```go
synthesized.NoNewPrivs = effectiveSc.AllowPrivilegeEscalation != nil && !*effectiveSc.AllowPrivilegeEscalation
```

WDYT?

Reply:
sounds good will update in a couple hours after meetings